ClassCastException NetworkPolicy

Hi, trying to add a network policy on an openshift client cause me some problems

Having a policy in a yaml file:

    apiVersion: networking.k8s.io/v1
    kind: NetworkPolicy
    metadata:
        name: some-name
    spec:
       xxxx

And trying to read the file using the client.load(..).get() method, gives me this error:

    java.lang.ClassCastException: io.fabric8.kubernetes.api.model.extensions.NetworkPolicy cannot be cast to io.fabric8.kubernetes.api.model.networking.NetworkPolicy
    at io.fabric8.kubernetes.client.handlers.networking.v1.NetworkPolicyHandler.edit(NetworkPolicyHandler.java:40)
    at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.acceptVisitors(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:332)
    at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.get(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:324)
    at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.get(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:65)

Not sure why there seems to be two implementations api.model.networking and api.model.extensions, 

I use kubernetes-client and openshift-client in version 4.10.0

Boiled it down to the serialization. Where the `client.network().networkPolicies().load(...)` sets the base type, the `client.load(...)` does not.

    @Test
    public void serializationWithType() {
        InputStream resourceAsStream = NetworkPolicyLoadBugTest.class.getResourceAsStream("/network/allow-namespace.yaml");
        Object unmarshal = Serialization.unmarshal(resourceAsStream, io.fabric8.kubernetes.api.model.networking.NetworkPolicy.class);
        Assert.assertEquals(io.fabric8.kubernetes.api.model.networking.NetworkPolicy.class, unmarshal.getClass());
    }

    @Test // fails
    public void serializationFromStream() {
        InputStream resourceAsStream = NetworkPolicyLoadBugTest.class.getResourceAsStream("/network/allow-namespace.yaml");
        Object unmarshal = Serialization.unmarshal(resourceAsStream);
        Assert.assertEquals(io.fabric8.kubernetes.api.model.networking.NetworkPolicy.class, unmarshal.getClass());
    }

Which results in the extension class instead

    java.lang.AssertionError: 
    Expected :class io.fabric8.kubernetes.api.model.networking.NetworkPolicy
    Actual   :class io.fabric8.kubernetes.api.model.extensions.NetworkPolicy

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,7 @@
 
 ### 4.10-SNAPSHOT
 #### Bugs
+* Fix #2251: Modify KubernetesDeserializer for handling classes with same name but different apiVersions 
 * Fix #2205: Event model classes from core v1 have been lost
 * Fix #2226: SharedIndexInformer for non-namespaced resources not working
 * Fix #2201: Uberjar doesn't contain model classes anymore

--- a/kubernetes-model/kubernetes-model-common/src/main/java/io/fabric8/kubernetes/model/util/Helper.java
+++ b/kubernetes-model/kubernetes-model-common/src/main/java/io/fabric8/kubernetes/model/util/Helper.java
@@ -15,8 +15,13 @@
  */
 package io.fabric8.kubernetes.model.util;
 
+import io.fabric8.kubernetes.model.annotation.ApiGroup;
+import io.fabric8.kubernetes.model.annotation.ApiVersion;
+
 import java.io.IOException;
 import java.io.InputStream;
+import java.lang.annotation.Annotation;
+import java.util.Arrays;
 import java.util.Scanner;
 
 public class Helper {
@@ -29,4 +34,22 @@ public class Helper {
             throw new RuntimeException(e);
         }
     }
+
+
+  public static String getAnnotationValue(Class kubernetesResourceType, Class annotationClass) {
+    Annotation annotation = getAnnotation(kubernetesResourceType, annotationClass);
+    if (annotation instanceof ApiGroup) {
+      return ((ApiGroup) annotation).value();
+    } else if (annotation instanceof ApiVersion) {
+      return ((ApiVersion) annotation).value();
+    }
+    return null;
+  }
+
+  private static Annotation getAnnotation(Class kubernetesResourceType, Class annotationClass) {
+    return Arrays.stream(kubernetesResourceType.getAnnotations())
+      .filter(annotation -> annotation.annotationType().equals(annotationClass))
+      .findFirst()
+      .orElse(null);
+  }
 }

--- a/kubernetes-model/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java
+++ b/kubernetes-model/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java
@@ -39,6 +39,9 @@ import io.fabric8.kubernetes.api.KubernetesResourceMappingProvider;
 import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
 import io.fabric8.kubernetes.api.model.KubernetesResource;
+import io.fabric8.kubernetes.model.annotation.ApiGroup;
+import io.fabric8.kubernetes.model.annotation.ApiVersion;
+import io.fabric8.kubernetes.model.util.Helper;
 
 public class KubernetesDeserializer extends JsonDeserializer<KubernetesResource> {
 
@@ -125,18 +128,26 @@ public class KubernetesDeserializer extends JsonDeserializer<KubernetesResource>
         private static final String KEY_SEPARATOR = "#";
         private static final String[] PACKAGES = {
                 "io.fabric8.kubernetes.api.model.",
-                "io.fabric8.kubernetes.api.model.admissionregistration.",
+                "io.fabric8.kubernetes.api.model.admission",
+                "io.fabric8.kubernetes.api.model.admissionregistration.v1.",
+                "io.fabric8.kubernetes.api.model.admissionregistration.v1beta1.",
                 "io.fabric8.kubernetes.api.model.apiextensions.",
                 "io.fabric8.kubernetes.api.model.apps.",
                 "io.fabric8.kubernetes.api.model.authentication.",
                 "io.fabric8.kubernetes.api.model.authorization.",
+                "io.fabric8.kubernetes.api.model.autoscaling.",
+                "io.fabric8.kubernetes.api.model.autoscaling.v1.",
+                "io.fabric8.kubernetes.api.model.autoscaling.v2beta1.",
+                "io.fabric8.kubernetes.api.model.autoscaling.v2beta2.",
                 "io.fabric8.kubernetes.api.model.batch.",
                 "io.fabric8.kubernetes.api.model.certificates.",
                 "io.fabric8.kubernetes.api.model.coordination.",
+                "io.fabric8.kubernetes.api.model.coordination.v1.",
                 "io.fabric8.kubernetes.api.model.discovery.",
                 "io.fabric8.kubernetes.api.model.extensions.",
                 "io.fabric8.kubernetes.api.model.events.",
                 "io.fabric8.kubernetes.api.model.networking.",
+                "io.fabric8.kubernetes.api.model.networking.v1beta1.",
                 "io.fabric8.kubernetes.api.model.policy.",
                 "io.fabric8.kubernetes.api.model.rbac.",
                 "io.fabric8.kubernetes.api.model.storage.",
@@ -159,7 +170,7 @@ public class KubernetesDeserializer extends JsonDeserializer<KubernetesResource>
         	if (clazz != null) {
         		return clazz; 
         	} 
-    		clazz = getInternalTypeForName(getClassName(key));
+    		clazz = getInternalTypeForName(key);
     		if (clazz != null) {
 				mappings.put(key, clazz);
 			}
@@ -221,15 +232,42 @@ public class KubernetesDeserializer extends JsonDeserializer<KubernetesResource>
             }
         }
 
-        private Class<? extends KubernetesResource> getInternalTypeForName(String name) {
+        private Class<? extends KubernetesResource> getInternalTypeForName(String key) {
+            String name = getClassName(key);
+            List<Class<? extends KubernetesResource>> possibleResults = new ArrayList<>();
+
+            // First pass, check if there are more than one class of same name
             for (String aPackage : PACKAGES) {
                 Class<? extends KubernetesResource> result = loadClassIfExists(aPackage + name);
                 if (result != null) {
+                    possibleResults.add(result);
+                }
+            }
+
+            // If only one class found, return it
+            if (possibleResults.size() == 1) {
+                return possibleResults.get(0);
+            }
+
+            // Compare with apiVersions being compared for set of classes found
+            for (Class<? extends KubernetesResource> result : possibleResults) {
+                String defaultKeyFromClass = getKeyFromClass(result);
+                if (key.equals(defaultKeyFromClass)) {
                     return result;
                 }
             }
             return null;
         }
+
+        private String getKeyFromClass(Class<? extends KubernetesResource> clazz) {
+          String apiGroup = Helper.getAnnotationValue(clazz, ApiGroup.class);
+          String apiVersion = Helper.getAnnotationValue(clazz, ApiVersion.class);
+          if (apiGroup != null && !apiGroup.isEmpty() && apiVersion != null && !apiVersion.isEmpty()) {
+            return createKey(apiGroup + "/" + apiVersion, clazz.getSimpleName());
+          }
+          return clazz.getSimpleName();
+        }
+
 
         private Class<? extends KubernetesResource> loadClassIfExists(String className) {
             try {

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DefaultSharedIndexInformerTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/DefaultSharedIndexInformerTest.java
@@ -214,6 +214,7 @@ class DefaultSharedIndexInformerTest {
     factory.startAllRegisteredInformers();
 
     relistSuccessful.await(LATCH_AWAIT_PERIOD_IN_SECONDS, TimeUnit.SECONDS);
+    waitUntilResourceVersionSynced();
     assertEquals(0, relistSuccessful.getCount());
     assertEquals(endResourceVersion, podInformer.lastSyncResourceVersion());
 

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/LoadTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/LoadTest.java
@@ -17,6 +17,7 @@
 package io.fabric8.kubernetes.client.mock;
 
 import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.networking.NetworkPolicy;
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.fabric8.kubernetes.client.server.mock.KubernetesServer;
 import org.junit.Rule;
@@ -83,5 +84,17 @@ public class LoadTest {
     assertEquals("v1", deploymentResource.getApiVersion());
     assertEquals("ImageStream", deploymentResource.getKind());
     assertEquals("eap-app", deploymentResource.getMetadata().getName());
+  }
+
+  @Test
+  void testNetworkPolicyLoad() {
+    KubernetesClient client = server.getClient();
+    List<HasMetadata> itemList = client.load(getClass().getResourceAsStream("/test-networkpolicy.yml")).get();
+
+    assertEquals(1, itemList.size());
+    NetworkPolicy ingress = (NetworkPolicy) itemList.get(0);
+    assertEquals("test-network-policy", ingress.getMetadata().getName());
+    assertEquals(1, ingress.getSpec().getIngress().size());
+    assertEquals(1, ingress.getSpec().getEgress().size());
   }
 }

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/NetworkingV1beta1IngressTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/NetworkingV1beta1IngressTest.java
@@ -15,6 +15,7 @@
  */
 package io.fabric8.kubernetes.client.mock;
 
+import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.networking.v1beta1.Ingress;
 import io.fabric8.kubernetes.api.model.networking.v1beta1.IngressBuilder;
 import io.fabric8.kubernetes.api.model.networking.v1beta1.IngressList;
@@ -28,6 +29,8 @@ import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.migrationsupport.rules.EnableRuleMigrationSupport;
 
+import java.util.List;
+
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
@@ -38,6 +41,17 @@ import static org.junit.jupiter.api.Assertions.assertTrue;
 public class NetworkingV1beta1IngressTest {
   @Rule
   public KubernetesServer server = new KubernetesServer();
+
+  @Test
+  void testLoad() {
+    KubernetesClient client = server.getClient();
+    List<HasMetadata> itemList = client.load(getClass().getResourceAsStream("/test-v1beta1-ingress.yml")).get();
+
+    assertEquals(1, itemList.size());
+    Ingress ingress = (Ingress) itemList.get(0);
+    assertEquals("tls-example-ingress", ingress.getMetadata().getName());
+    assertEquals(1, ingress.getSpec().getTls().size());
+  }
 
   @Test
   public void testList() {

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/V1HorizontalPodAutoscalerTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/V1HorizontalPodAutoscalerTest.java
@@ -16,6 +16,7 @@
 
 package io.fabric8.kubernetes.client.mock;
 
+import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.autoscaling.v1.HorizontalPodAutoscaler;
 import io.fabric8.kubernetes.api.model.autoscaling.v1.HorizontalPodAutoscalerList;
 import io.fabric8.kubernetes.api.model.autoscaling.v1.HorizontalPodAutoscalerBuilder;
@@ -29,6 +30,8 @@ import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.migrationsupport.rules.EnableRuleMigrationSupport;
 
+import java.util.List;
+
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
@@ -39,6 +42,19 @@ import static org.junit.jupiter.api.Assertions.assertFalse;
 public class V1HorizontalPodAutoscalerTest {
   @Rule
   public KubernetesServer server = new KubernetesServer();
+
+  @Test
+  void testLoadAsPlainList() {
+    KubernetesClient client = server.getClient();
+    List<HasMetadata> itemsList = client.load(getClass().getResourceAsStream("/test-horizontalpodautoscaler.yml")).get();
+
+    assertNotNull(itemsList);
+    assertEquals(1, itemsList.size());
+    HorizontalPodAutoscaler horizontalPodAutoscaler = (HorizontalPodAutoscaler) itemsList.get(0);
+    assertEquals("php-apache", horizontalPodAutoscaler.getMetadata().getName());
+    assertEquals(10, horizontalPodAutoscaler.getSpec().getMaxReplicas());
+    assertEquals(1, horizontalPodAutoscaler.getSpec().getMinReplicas());
+  }
 
   @Test
   public void testList() {

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/V2beta2HorizontalPodAutoscalerTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/V2beta2HorizontalPodAutoscalerTest.java
@@ -16,6 +16,7 @@
 
 package io.fabric8.kubernetes.client.mock;
 
+import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.autoscaling.v2beta2.HorizontalPodAutoscaler;
 import io.fabric8.kubernetes.api.model.autoscaling.v2beta2.HorizontalPodAutoscalerList;
 import io.fabric8.kubernetes.api.model.autoscaling.v2beta2.HorizontalPodAutoscalerBuilder;
@@ -29,6 +30,8 @@ import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.migrationsupport.rules.EnableRuleMigrationSupport;
 
+import java.util.List;
+
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
@@ -39,6 +42,18 @@ import static org.junit.jupiter.api.Assertions.assertFalse;
 public class V2beta2HorizontalPodAutoscalerTest {
   @Rule
   public KubernetesServer server = new KubernetesServer();
+
+  @Test
+  void testLoadAsPlainList() {
+    KubernetesClient client = server.getClient();
+    List<HasMetadata> itemsList = client.load(getClass().getResourceAsStream("/test-v2beta2-hpa.yml")).get();
+
+    assertNotNull(itemsList);
+    assertEquals(1, itemsList.size());
+    HorizontalPodAutoscaler horizontalPodAutoscaler = (HorizontalPodAutoscaler) itemsList.get(0);
+    assertEquals("php-apache", horizontalPodAutoscaler.getMetadata().getName());
+    assertEquals(1, horizontalPodAutoscaler.getSpec().getMetrics().size());
+  }
 
   @Test
   public void testList() {

--- a/kubernetes-tests/src/test/resources/test-networkpolicy.yml
+++ b/kubernetes-tests/src/test/resources/test-networkpolicy.yml
@@ -1,0 +1,50 @@
+#
+# Copyright (C) 2015 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: test-network-policy
+  namespace: default
+spec:
+  podSelector:
+    matchLabels:
+      role: db
+  policyTypes:
+    - Ingress
+    - Egress
+  ingress:
+    - from:
+        - ipBlock:
+            cidr: 172.17.0.0/16
+            except:
+              - 172.17.1.0/24
+        - namespaceSelector:
+            matchLabels:
+              project: myproject
+        - podSelector:
+            matchLabels:
+              role: frontend
+      ports:
+        - protocol: TCP
+          port: 6379
+  egress:
+    - to:
+        - ipBlock:
+            cidr: 10.0.0.0/24
+      ports:
+        - protocol: TCP
+          port: 5978

--- a/kubernetes-tests/src/test/resources/test-v1beta1-ingress.yml
+++ b/kubernetes-tests/src/test/resources/test-v1beta1-ingress.yml
@@ -1,0 +1,33 @@
+#
+# Copyright (C) 2015 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+apiVersion: networking.k8s.io/v1beta1
+kind: Ingress
+metadata:
+  name: tls-example-ingress
+spec:
+  tls:
+    - hosts:
+        - sslexample.foo.com
+      secretName: testsecret-tls
+  rules:
+    - host: sslexample.foo.com
+      http:
+        paths:
+          - path: /
+            backend:
+              serviceName: service1
+              servicePort: 80

--- a/kubernetes-tests/src/test/resources/test-v2beta2-hpa.yml
+++ b/kubernetes-tests/src/test/resources/test-v2beta2-hpa.yml
@@ -1,0 +1,34 @@
+#
+# Copyright (C) 2015 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+apiVersion: autoscaling/v2beta2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: php-apache
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: php-apache
+  minReplicas: 1
+  maxReplicas: 10
+  metrics:
+    - type: Resource
+      resource:
+        name: cpu
+        target:
+          type: Utilization
+          averageUtilization: 50

